cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java

```diff
@@ -0,0 +1,177 @@
+ * the License.
+ */
+
+package co.cask.cdap.etl.batch.spark;
+
+import co.cask.cdap.api.data.batch.OutputFormatProvider;
+import co.cask.cdap.api.spark.SparkContext;
+import com.google.common.collect.ImmutableMap;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.spark.api.java.JavaPairRDD;
+
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *
+ */
+final class SparkBatchSinkFactory {
+
+  static SparkBatchSinkFactory deserialize(InputStream inputStream) throws IOException {
+    DataInput input = new DataInputStream(inputStream);
+    Map<String, OutputFormatProvider> outputFormatProviders = Serializations.deserializeMap(
+      input, new Serializations.ObjectReader<OutputFormatProvider>() {
+      @Override
+      public OutputFormatProvider read(DataInput input) throws IOException {
+        return new BasicOutputFormatProvider(input.readUTF(),
+                                             Serializations.deserializeMap(input,
+                                                                           Serializations.createStringObjectReader()));
+      }
+    });
+    Map<String, DatasetInfo> datasetInfos = Serializations.deserializeMap(
+      input, new Serializations.ObjectReader<DatasetInfo>() {
+      @Override
+      public DatasetInfo read(DataInput input) throws IOException {
+        return DatasetInfo.deserialize(input);
+      }
+    });
+    Map<String, Set<String>> sinkOutputs = Serializations.deserializeMap(
+      input, Serializations.createStringSetObjectReader());
+    return new SparkBatchSinkFactory(outputFormatProviders, datasetInfos, sinkOutputs);
+  }
+
+  private final Map<String, OutputFormatProvider> outputFormatProviders;
+  private final Map<String, DatasetInfo> datasetInfos;
+  private final Map<String, Set<String>> sinkOutputs;
+
+  SparkBatchSinkFactory() {
+    this.outputFormatProviders = new HashMap<>();
+    this.datasetInfos = new HashMap<>();
+    this.sinkOutputs = new HashMap<>();
+  }
+
+  private SparkBatchSinkFactory(Map<String, OutputFormatProvider> providers,
+                                Map<String, DatasetInfo> datasetInfos,
+                                Map<String, Set<String>> sinkOutputs) {
+    this.outputFormatProviders = providers;
+    this.datasetInfos = datasetInfos;
+    this.sinkOutputs = sinkOutputs;
+  }
+
+  void addOutput(String stageName, String outputName, OutputFormatProvider outputFormatProvider) {
+    if (outputFormatProviders.containsKey(outputName) || datasetInfos.containsKey(outputName)) {
+      throw new IllegalArgumentException("OutputFormatProvider with name " + outputName + " is already defined");
```

Comment:
Please change to user friendly message. `Cannot add multiple outputs with the same name` or something to that effect